File: smoke_tests/csharp/authentication.cs
// Authentication vulnerabilities in C#
using System;
using System.Web;
using System.Web.Mvc;
using System.Web.Security;
using System.Security.Cryptography;
using System.Text;

namespace VulnerableApp
{
    public class AuthController : Controller
    {
        // Test 1: Plaintext password storage
        public ActionResult Register()
        {
            string username = Request.Form["username"];
            string password = Request.Form["password"];
            // VULNERABLE: Storing plaintext password
            SaveUser(username, password);
            return Ok();
        }

        // Test 2: Weak password hashing (MD5)
        public ActionResult RegisterMd5()
        {
            string password = Request.Form["password"];
            // VULNERABLE: MD5 is too weak for passwords
            using (var md5 = MD5.Create())
            {
                byte[] hash = md5.ComputeHash(Encoding.UTF8.GetBytes(password));
                string hashedPassword = BitConverter.ToString(hash).Replace("-", "");
                SaveUser(Request.Form["username"], hashedPassword);
            }
            return Ok();
        }

        // Test 3: Unsalted SHA hash
        public ActionResult RegisterSha()
        {
            string password = Request.Form["password"];
            // VULNERABLE: No salt, allows rainbow table attacks
            using (var sha = SHA256.Create())
            {
                byte[] hash = sha.ComputeHash(Encoding.UTF8.GetBytes(password));
                SaveUser(Request.Form["username"], Convert.ToBase64String(hash));
            }
            return Ok();
        }

        // Test 4: Hardcoded credentials
        public ActionResult AdminLogin()
        {
            string username = Request.Form["username"];
            string password = Request.Form["password"];
            // VULNERABLE: Hardcoded credentials
            if (username == "admin" && password == "admin123")
            {
                FormsAuthentication.SetAuthCookie(username, false);
                return RedirectToAction("Dashboard");
            }
            return View();
        }

        // Test 5: Timing attack in comparison
        public ActionResult Login()
        {
            string username = Request.Form["username"];
            string password = Request.Form["password"];
            string storedHash = GetPasswordHash(username);
            string inputHash = HashPassword(password);
            // VULNERABLE: String comparison leaks timing information
            if (storedHash == inputHash)
            {
                return Ok();
            }
            return Unauthorized();
        }

        // Test 6: No account lockout
        public ActionResult LoginNoLockout()
        {
            string username = Request.Form["username"];
            string password = Request.Form["password"];
            // VULNERABLE: No failed attempt tracking/lockout
            if (ValidateCredentials(username, password))
            {
                return Ok();
            }
            return Unauthorized();
        }

        // Test 7: Password in URL
        public ActionResult LoginViaGet(string username, string password)
        {
            // VULNERABLE: Credentials in URL (logged, cached, referer header)
            if (ValidateCredentials(username, password))
            {
                return Ok();
            }
            return Unauthorized();
        }

        // Test 8: Password logged
        public ActionResult LoginWithLogging()
        {
            string username = Request.Form["username"];
            string password = Request.Form["password"];
            // VULNERABLE: Password in logs
            System.Diagnostics.Debug.WriteLine($"Login attempt: {username}/{password}");
            return Ok();
        }

        // Test 9: Weak session configuration
        public ActionResult CreateSession()
        {
            // VULNERABLE: Session cookie without secure flags
            Response.Cookies.Add(new HttpCookie("sessionId", Guid.NewGuid().ToString())
            {
                HttpOnly = false,  // Vulnerable to XSS
                Secure = false     // Sent over HTTP
            });
            return Ok();
        }

        // Test 10: Insufficient password requirements
        public ActionResult SetPassword()
        {
            string password = Request.Form["password"];
            // VULNERABLE: No complexity requirements
            if (password.Length >= 4)  // Too short, no complexity check
            {
                SavePassword(password);
            }
            return Ok();
        }

        // Test 11: Remember me with long expiration
        public ActionResult PersistentLogin()
        {
            string username = Request.Form["username"];
            // VULNERABLE: Very long token lifetime
            FormsAuthentication.SetAuthCookie(username, true);
            var cookie = Response.Cookies[FormsAuthentication.FormsCookieName];
            cookie.Expires = DateTime.Now.AddYears(10);  // Too long
            return Ok();
        }

        // Placeholder methods
        private void SaveUser(string username, string password) { }
        private string GetPasswordHash(string username) => "";
        private string HashPassword(string password) => "";
        private bool ValidateCredentials(string u, string p) => false;
        private void SavePassword(string p) { }
    }
}

File: smoke_tests/php/authentication.php
<?php
// Authentication vulnerabilities in PHP

// Test 1: Plaintext password storage
function register_plaintext() {
    $username = $_POST['username'];
    $password = $_POST['password'];

    $pdo = get_pdo();
    // VULNERABLE: Storing plaintext password
    $stmt = $pdo->prepare("INSERT INTO users (username, password) VALUES (?, ?)");
    $stmt->execute([$username, $password]);
}

// Test 2: MD5 password hashing
function register_md5() {
    $password = $_POST['password'];
    // VULNERABLE: MD5 is too weak
    $hash = md5($password);
    store_user($_POST['username'], $hash);
}

// Test 3: SHA1 without salt
function register_sha1() {
    $password = $_POST['password'];
    // VULNERABLE: Unsalted SHA1
    $hash = sha1($password);
    store_user($_POST['username'], $hash);
}

// Test 4: Hardcoded credentials
function admin_login() {
    $username = $_POST['username'];
    $password = $_POST['password'];
    // VULNERABLE: Hardcoded credentials
    if ($username === 'admin' && $password === 'admin123') {
        $_SESSION['admin'] = true;
        return true;
    }
    return false;
}

// Test 5: Timing attack in comparison
function login_timing() {
    $username = $_POST['username'];
    $password = $_POST['password'];

    $user = get_user_by_username($username);
    // VULNERABLE: String comparison leaks timing
    if ($user && $user['password_hash'] === hash('sha256', $password)) {
        $_SESSION['user_id'] = $user['id'];
        return true;
    }
    return false;
}

// Test 6: No account lockout
function login_no_lockout() {
    $username = $_POST['username'];
    $password = $_POST['password'];

    // VULNERABLE: No failed attempt tracking
    $user = get_user_by_username($username);
    if ($user && password_verify($password, $user['password_hash'])) {
        $_SESSION['user_id'] = $user['id'];
        return true;
    }
    return false;
}

// Test 7: Password in URL (GET request)
function login_get() {
    $username = $_GET['username'];
    $password = $_GET['password'];
    // VULNERABLE: GET request with credentials
    return authenticate($username, $password);
}

// Test 8: Password logged
function login_with_logging() {
    $username = $_POST['username'];
    $password = $_POST['password'];
    // VULNERABLE: Password in logs
    error_log("Login attempt: $username/$password");
    return authenticate($username, $password);
}

// Test 9: Weak session configuration
function create_session() {
    // VULNERABLE: Insecure session settings
    ini_set('session.cookie_httponly', 0);
    ini_set('session.cookie_secure', 0);
    session_start();
    $_SESSION['user_id'] = get_user_id();
}

// Test 10: Insufficient password requirements
function set_password() {
    $password = $_POST['password'];
    // VULNERABLE: No complexity check
    if (strlen($password) >= 4) {  // Too short
        update_password($password);
        return true;
    }
    return false;
}

// Test 11: Predictable remember token
function remember_me() {
    $user_id = $_SESSION['user_id'];
    // VULNERABLE: Predictable remember token
    $token = $user_id . time();
    setcookie('remember_token', $token, time() + 86400 * 365);
}

// Test 12: Session fixation
function login_no_regenerate() {
    if (authenticate($_POST['username'], $_POST['password'])) {
        // VULNERABLE: Session not regenerated after login
        $_SESSION['user_id'] = get_user_id();
        return true;
    }
    return false;
}

// Helper functions
function get_pdo() { return new PDO('sqlite::memory:'); }
function store_user($username, $hash) {}
function get_user_by_username($username) { return null; }
function authenticate($username, $password) { return false; }
function get_user_id() { return 1; }
function update_password($password) {}
?>
